### DISCUSSION TITLE
Using SDL for the plugin window

### DISCUSSION THREAD

**Discussion by Photosounder** [iPlug2/iPlug2#1066] [Q&A] [answered]
After using WDL-OL in 2011-2015 for my commercial plugins I'm back with a more low level need than before: this time not only do I still need to display a framebuffer in a window, but I also need all inputs given to the plugin window all at once. Can I and should I use SDL for this, is there any downside to using SDL, like is there anything wrong with making a SDL window (I don't know much about how DAWs handle windows) instead of whatever IPlug2 uses? Like would it cause integration problem with the DAW, or would the DAW have problems with control automation? I see that the two IPlug2 [graphic backends](https://github.com/iPlug2/iPlug2/wiki/IGraphics-backend-comparison) have their issues so maybe using SDL would be a way to sidestep their issues as well.

If you want to know the details of why I might need SDL instead of what's currently offered it's because I need to:

- Get all keyboard presses in the window as both scancodes and characters
- Get raw text input, including pasting text from the clipboard, but without a visible control, I just need to get strings of characters as they come (I guess individual character entry through key presses and clipboard pasting could have two totally different solutions)
- Be able to copy text to the clipboard
- Get all mouse inputs in the window, that means position relative to the window, relative movement at all times and button and wheel events
- Be able to selectively capture the mouse cursor, that means make it disappear while it stays in the window no matter what and continues to report relative movement and button states
- Be able to set the mouse cursor's position, this is mostly important when uncapturing the mouse cursor to decide where it should pop back up (typically it needs to pop back up either in the middle of the window or right where it was when it was captured).

I need something so radically low level and generic because I'm writing an open source host plugin for my own kind of module (WebAssembly modules that communicate with their host through text, see host library [here](https://github.com/Photosounder/WAHE)), and those modules do their own GUI in an isolated and independent way that the host knows nothing about, they just need the host to provide all the low level inputs, control over the mouse and access to the clipboard. For instance a module will have its own text editor controls, but it's impossible to know anything about the text editing outside of the module, from the outside all we see is pixels. Likewise a module might capture the mouse but we don't know what for, it could be for moving a knob indefinitely, or it could be for playing Doom, who knows.

**Comment by olilarkin** [comment, marked as answer]
I don't really know the practical implications of using SDL in a plug-in. @madronalabs is doing something with his framework here: https://github.com/madronalabs/mlvg . IIRC GLFW is difficult to use in a plug-in, I am not sure if SDL has the same issue. [PUGL](https://github.com/lv2/pugl) is like GLFW but designed of plug-ins, and looks pretty good. You can find a WIP example of it being used in iPlug2 [here](https://github.com/iPlug2/iPlug2/tree/webview/platform-dialogs/Examples/IPlugOpenGL). [CPLUG](https://github.com/Tremus/CPLUG) looks like a nice new plugin abstraction that might be worth investigating for your project

**Reply by Photosounder** [reply]
Thanks Oli, CPLUG might indeed be more my speed.
